src/Database/LSMTree/Internal/Run.hs

```diff
@@ -30,4 +32,115 @@
+module Database.LSMTree.Internal.Run (
+    -- * Paths
+    module FsPaths
+    -- * Run
+  , RefCount
+  , Run
+  , addReference
+  , removeReference
+  , fromWriteBuffer
+  ) where
+
+import           Control.Monad (when)
+import           Data.Foldable (for_)
+import           Data.IORef
+import           Database.LSMTree.Internal.Entry
+import           Database.LSMTree.Internal.Run.BloomFilter (Bloom)
+import           Database.LSMTree.Internal.Run.FsPaths as FsPaths
+import           Database.LSMTree.Internal.Run.Index.Compact (CompactIndex)
+import           Database.LSMTree.Internal.Run.Mutable
+import           Database.LSMTree.Internal.Serialise
+import           Database.LSMTree.Internal.WriteBuffer (WriteBuffer)
+import qualified Database.LSMTree.Internal.WriteBuffer as WB
+import qualified System.FS.API as FS
+import           System.FS.API (HasFS)
+
+
+-- | The in-memory representation of a completed LSM run.
+--
+data Run fhandle = Run {
+       -- | The reference count for the LSM run. This counts the
+       -- number of references from LSM handles to this run. When
+       -- this drops to zero the open files will be closed.
+       lsmRunRefCount :: !RefCount,
+
+       -- | The file system paths for all the files used by the run.
+       lsmRunFsPaths  :: !RunFsPaths,
+
+       -- | The bloom filter for the set of keys in this run.
+       lsmRunFilter   :: !(Bloom SerialisedKey),
+
+       -- | The in-memory index mapping keys to page numbers in the
+       -- Key\/Ops file. In future we may support alternative index
+       -- representations.
+       lsmRunIndex    :: !CompactIndex,
+
+       -- | The file handle for the Key\/Ops file. This file is opened
+       -- read-only and is accessed in a page-oriented way, i.e. only
+       -- reading whole pages, at page offsets. It will be opened with
+       -- 'O_DIRECT' on supported platforms.
+       lsmRunKOpsFile :: !fhandle,
+
+       -- | The file handle for the BLOBs file. This file is opened
+       -- read-only and is accessed in a normal style using buffered
+       -- I\/O, reading arbitrary file offset and length spans.
+       lsmRunBlobFile :: !fhandle
+     }
+
+addReference :: HasFS IO h -> Run (FS.Handle h) -> IO ()
+addReference _ Run {..} =
+    atomicModifyIORef lsmRunRefCount (\n -> (n+1, ()))
+
+-- | After calling this operation, the run must not be used anymore.
+-- Might cause the run to be closed, which removes all its associated
+-- files from disk.
+removeReference :: HasFS IO h -> Run (FS.Handle h) -> IO ()
+removeReference fs run@Run {..} = do
+    count <- atomicModifyIORef' lsmRunRefCount (\n -> (n-1, n-1))
+    when (count <= 0) $
+      close fs run
+
+-- | After calling this operation, the run must not be used anymore.
+-- Removes all files associated with the run from disk.
+close :: HasFS IO h -> Run (FS.Handle h) -> IO ()
+close fs Run {..} = do
+    -- close and remove all files
+    -- TODO: even if one of them throws an exception
```

Comment:
We can test this using `fs-sim`. The scope of this PR is probably already too large to test resource cleanup, but we should at some point. Maybe add a GH issue before merging this PR?

Reply:
Yes, that is a good idea and I'll do it in a follow up PR.